compiler/rustc_hir_analysis/src/autoderef.rs

```diff
@@ -68,7 +68,14 @@ impl<'a, 'tcx> Iterator for Autoderef<'a, 'tcx> {
             return None;
         }
 
-        if self.state.cur_ty.is_ty_var() {
+        // We want to support method and function calls for `impl Deref<Target = ..>`.
+        //
+        // To do so we don't eagerly bail if the current type is the hidden type of an
+        // opaque type and instead return `None` in `fn overloaded_deref_ty` if the
+        // opaque does not have a `Deref` item-bound.
+        if let &ty::Infer(ty::TyVar(vid)) = self.state.cur_ty.kind()
```

Comment:
similar here, I realise we previously only checked `is_ty_var`, but how does the rest of this codepath handle int/float vars

Reply:
they don't implement `Deref` so they are the final self type and users of autoderef have to handle them if necessary